src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DocumentState.cs

```diff
@@ -193,98 +193,113 @@ public virtual DocumentState WithTextLoader(TextLoader textLoader)
     {
-        // OK we have to generate the code.
-        using var importSources = new PooledArrayBuilder<RazorSourceDocument>(imports.Length);
-        foreach (var item in imports)
-        {
-            var importProjectItem = item.FilePath is null ? null : projectEngine.FileSystem.GetItem(item.FilePath, item.FileKind);
-            var sourceDocument = await GetRazorSourceDocumentAsync(item.Document, importProjectItem, cancellationToken).ConfigureAwait(false);
-            importSources.Add(sourceDocument);
-        }
+        using var importSources = new PooledArrayBuilder<RazorSourceDocument>(importItems.Length);
 
-        var projectItem = document.FilePath is null ? null : projectEngine.FileSystem.GetItem(document.FilePath, document.FileKind);
-        var documentSource = await GetRazorSourceDocumentAsync(document, projectItem, cancellationToken).ConfigureAwait(false);
-
-        if (forceRuntimeCodeGeneration)
+        foreach (var importItem in importItems)
         {
-            return projectEngine.Process(documentSource, fileKind: document.FileKind, importSources.DrainToImmutable(), tagHelpers);
-        }
+            var importProjectItem = importItem is { FilePath: string filePath, FileKind: var fileKind }
+                ? projectEngine.FileSystem.GetItem(filePath, fileKind)
+                : null;
 
-        return projectEngine.ProcessDesignTime(documentSource, fileKind: document.FileKind, importSources.DrainToImmutable(), tagHelpers);
-    }
-
-    internal static async Task<ImmutableArray<ImportItem>> GetImportsAsync(IDocumentSnapshot document, RazorProjectEngine projectEngine, CancellationToken cancellationToken)
-    {
-        var imports = GetImportsCore(document.Project, projectEngine, document.FilePath.AssumeNotNull(), document.FileKind.AssumeNotNull());
-        using var result = new PooledArrayBuilder<ImportItem>(imports.Length);
+            var properties = RazorSourceDocumentProperties.Create(importItem.FilePath, importProjectItem?.RelativePhysicalPath);
+            var importSource = RazorSourceDocument.Create(importItem.Text, properties);
 
-        foreach (var snapshot in imports)
-        {
-            var versionStamp = await snapshot.GetTextVersionAsync(cancellationToken).ConfigureAwait(false);
-            result.Add(new ImportItem(snapshot.FilePath, versionStamp, snapshot));
+            importSources.Add(importSource);
         }
 
-        return result.DrainToImmutable();
+        return importSources.DrainToImmutable();
     }
 
-    private static async Task<RazorSourceDocument> GetRazorSourceDocumentAsync(
-        IDocumentSnapshot document,
-        RazorProjectItem? projectItem,
-        CancellationToken cancellationToken)
+    private static async Task<RazorSourceDocument> GetSourceAsync(IDocumentSnapshot document, RazorProjectEngine projectEngine, CancellationToken cancellationToken)
```

Comment:
Might not be worth it, but if this method took `string filePath, string fileKind, SourceText text`, then it could be called from inside the loop in `GetImportSources` too (and become sync.

Reply:
I'll definitely look at that in my next PR in this space.